crates/sui-indexer/src/type_tag_parser.rs

```diff
@@ -0,0 +1,470 @@
+// Copyright (c) Mysten Labs, Inc.
+// Copyright (c) The Diem Core Contributors
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, format_err, Result};
+use fastcrypto::encoding::{Encoding, Hex};
+use move_core_types::{
+    account_address::AccountAddress,
+    identifier::{self, Identifier},
+    language_storage::{StructTag, TypeTag},
+};
+use std::iter::Peekable;
+
+#[derive(Eq, PartialEq, Debug)]
+enum Token {
+    U8Type,
+    U16Type,
+    U32Type,
+    U64Type,
+    U128Type,
+    U256Type,
+    BoolType,
+    AddressType,
+    VectorType,
+    SignerType,
+    Whitespace(String),
+    Name(String),
+    Address(String),
+    U8(String),
+    U16(String),
+    U32(String),
+    U64(String),
+    U128(String),
+    U256(String),
+
+    Bytes(String),
+    True,
+    False,
+    ColonColon,
+    Lt,
+    Gt,
+    Comma,
+    Eof,
+}
+
+impl Token {
+    fn is_whitespace(&self) -> bool {
+        matches!(self, Self::Whitespace(_))
+    }
+}
+
+fn token_to_name(t: Token) -> Result<String> {
+    Ok(match t {
+        Token::U8Type => "u8".to_string(),
+        Token::U16Type => "u16".to_string(),
+        Token::U32Type => "u32".to_string(),
+        Token::U64Type => "u64".to_string(),
+        Token::U128Type => "u128".to_string(),
+        Token::U256Type => "u256".to_string(),
+        Token::BoolType => "bool".to_string(),
+        Token::AddressType => "address".to_string(),
+        Token::VectorType => "vector".to_string(),
+        Token::True => "true".to_string(),
+        Token::False => "false".to_string(),
+        Token::SignerType => "signer".to_string(),
+        Token::Name(name) => name,
+        _ => bail!("Unexpected token {:?} expected a name", t),
+    })
+}
+
+fn name_token(s: String) -> Token {
+    match s.as_str() {
+        "u8" => Token::U8Type,
+        "u16" => Token::U16Type,
+        "u32" => Token::U32Type,
+        "u64" => Token::U64Type,
+        "u128" => Token::U128Type,
+        "u256" => Token::U256Type,
+        "bool" => Token::BoolType,
+        "address" => Token::AddressType,
+        "vector" => Token::VectorType,
+        "true" => Token::True,
+        "false" => Token::False,
+        "signer" => Token::SignerType,
+        _ => Token::Name(s),
+    }
+}
+
+fn next_number(initial: char, mut it: impl Iterator<Item = char>) -> Result<(Token, usize)> {
+    let mut num = String::new();
+    num.push(initial);
+    loop {
+        match it.next() {
+            Some(c) if c.is_ascii_digit() || c == '_' => num.push(c),
+            Some(c) if c.is_alphanumeric() => {
```

Comment:
should it be ascii? also if this is just looking for a `u` why not match on that only?